src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -315,6 +322,10 @@ entails SolverOptions{..} constraint context hints =
             -- Make a dictionary from subgoal dictionaries by applying the correct function
             mkDictionary :: Evidence -> Maybe [Expr] -> m Expr
             mkDictionary (NamedInstance n) args = return $ foldl App (Var n) (fold args)
+            mkDictionary UnionInstance (Just [e]) =
+              -- We need the subgoal dictionary to appear in the term somewhere
+              return $ App (Abs (VarBinder (Ident C.__unused)) (Literal (ObjectLiteral []))) e
+            mkDictionary UnionInstance _ = return $ Literal (ObjectLiteral [])
```

Comment:
I think we should be using `C.undefined` values for all of these; like `C.__unused` it gets removed during codegen... we introduced it to remove the `ObjectLiteral []` applications we used to have everywhere, as they add up to be a significant performance penalty.